### DISCUSSION TITLE
Why I can't use avmplus.describeType?

### DISCUSSION THREAD

**Discussion by ylazy** [airsdk/Adobe-Runtime-Support#2285] [Q&A] [answered]
when I try to use:
```
import avmplus.*;
trace((avmplus.describeType(object, avmplus.INCLUDE_BASES|avmplus.INCLUDE_INTERFACES|avmplus.INCLUDE_VARIABLES|avmplus.INCLUDE_ACCESSORS|avmplus.INCLUDE_METHODS|avmplus.INCLUDE_METADATA|avmplus.INCLUDE_CONSTRUCTOR|avmplus.INCLUDE_TRAITS|avmplus.USE_ITRAITS) as XML).toXMLString());
```
The compiler display error: Cannot access a property or method of a null object reference.
Why I can't use the avmplus.describeType method? It's located in airglobal.swc

**Comment by ajwfrost** [comment, marked as answer]
I think that may be a special compiler setting that you'd need to access private/internal avmplus stuff ... but the flash.utils.describeType method (and other similar functions) just call straight through to the avmplus equivalents.

**Reply by ylazy** [reply]
the flash.utils.describeType method does not have the second argument (flags).
It seems to be the avmplus.describeType is a public method.
Could you give me more info about the compiler setting that you said?
Thanks!

**Reply by ajwfrost** [reply]
I thought I'd seen an argument like `-avmplus` but just looking at it now, when we build the airglobal.swc file, we've got `-AS3` and also `-builtin` .. but it's not the normal AS3 compiler, I think it may be an older one (asc.jar? I think it's available from Flex?)

Will have a look though, I'm not entirely sure why/how it's set as private..!

**Reply by ajwfrost** [reply]
FYI - not quite sure why it doesn't work unless it's related to the syntax, but you can actually call it manually via reflection:
```
			var fn : Function = getDefinitionByName("avmplus::describeType");
			var xmlResult : XML = fn.call(null, this, 511);
			trace(xmlResult);
```

**Reply by ylazy** [reply]
It worked! Thank u so much!!!

**Reply by ylazy** [reply]
replace 511 by 8 or 8|32, the error occurred again :D

**Reply by ajwfrost** [reply]
Oh okay, now I see what's happening .. some of the flags will end up with invalid values, because the first thing that happens internally is to describe the type using those flags - and sometimes that can end up with a null object. But there's then also some extra processing that happens in the avmplus method, hence there is a possibility that you get a null dereference internally within that function.

It's meant to be called with the "FLASH10" compatibility setting which is a bunch of the flags = 0x5FF (1535 decimal)

**Reply by pnico** [reply]
We used to use this, we were also able to get these functions to work without getDefinitionByName by creating our own 'avmplus' package/folder - in **our** project 😛 - with some class like "MyAvmTools.as" in it, somehow it's ok if we use our own class added to the avmplus package from airglobal.swc and then call avmplus functions from within that class..
```
package avmplus {
	public class MyAvmTools {
		public static function myDescribeTypeJson( obj:*, flags:uint ) : Object {
			return describeTypeJSON( obj, flags );
		}
	}
}
```
It was a long time ago so I don't know if this still works.. a little bit nicer though maybe.
Edit: actually here this looks better: https://www.jacksondunstan.com/articles/2613

**Reply by ylazy** [reply]
Thank you all! You all helped me a lot.
@pnico Thank you for the link. It worked! Tested with all flags.
